// @pbkit/runtime/rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createServerImplBuilder = exports.getMethodImpl = void 0;
const event_buffer_1 = require("./async/event-buffer");
const observer_1 = require("./async/observer");
function getMethodImpl(handler) {
    return (messages, metadata) => {
        const headerPromise = (0, observer_1.defer)();
        const trailerPromise = (0, observer_1.defer)();
        const drainEnd = (0, observer_1.defer)();
        const eventBuffer = (0, event_buffer_1.createEventBuffer)({
            onDrainEnd: drainEnd.resolve,
        });
        const header = headerPromise.resolve;
        const send = eventBuffer.push;
        const end = (value) => {
            eventBuffer.finish();
            trailerPromise.resolve(value);
        };
        handler({ metadata, messages, drainEnd }, { header, send, end });
        return [eventBuffer.drain(), headerPromise, trailerPromise];
    };
}
exports.getMethodImpl = getMethodImpl;
function createServerImplBuilder() {
    const buffer = (0, event_buffer_1.createEventBuffer)();
    return {
        register(methodDescriptor, handler) {
            buffer.push([methodDescriptor, getMethodImpl(handler)]);
        },
        finish: buffer.finish,
        drain: buffer.drain,
    };
}
exports.createServerImplBuilder = createServerImplBuilder;
